aptos-move/block-executor/src/view.rs

```diff
@@ -158,11 +167,149 @@ impl<'a, T: Transaction, X: Executable> ParallelState<'a, T, X> {
+            DependencyResult::Resolved => true,
+        }
+    }
+
+    fn get_resource_from_group(
+        &self,
+        txn_idx: TxnIndex,
+        key: &T::Key,
+        tag: &T::Tag,
+        return_group_size: bool,
+        check_existence: bool,
+    ) -> GroupReadResult {
+        if check_existence {
+            assert!(
+                !return_group_size,
+                "Group size should not be requested when checking existence"
+            );
+        }
+
+        loop {
+            match self.versioned_map.group_data().read_from_group(
+                key,
+                txn_idx,
+                tag,
+                return_group_size,
+            ) {
+                Ok((res, group_size, maybe_version)) => {
+                    if return_group_size {
+                        // Register a read to make sure size of a group is validated.
+                        // (gas charging may depend on it).
+                        self.captured_reads
+                            .borrow_mut()
+                            .push(ReadDescriptor::from_group_size(
+                                key.clone(),
+                                group_size.expect("group size must be returned"),
+                            ));
+                    } else {
+                        assert_none!(group_size, "not asked to return group size");
+                    }
+
+                    if check_existence {
+                        // Only record exitence query.
+                        self.captured_reads
+                            .borrow_mut()
+                            .push(ReadDescriptor::from_exists(
+                                key.clone(),
+                                Some(tag.clone()),
+                                !res.is_deletion(),
+                            ));
+                    } else {
+                        match maybe_version {
+                            // Record the versioned read.
+                            Some((idx, incarnation)) => {
+                                self.captured_reads
+                                    .borrow_mut()
+                                    .push(ReadDescriptor::from_version(
+                                        key.clone(),
+                                        Some(tag.clone()),
+                                        idx,
+                                        incarnation,
+                                    ))
+                            },
+                            None => self
+                                .captured_reads
+                                .borrow_mut()
+                                .push(ReadDescriptor::from_storage(key.clone(), Some(tag.clone()))),
+                        }
+                    }
+
+                    return GroupReadResult::Value((res.extract_raw_bytes(), group_size));
+                },
+                Err(MVGroupError::NotFound) => {
+                    if check_existence {
```

Comment:
Need to record in any case

Reply:
What if we wrapped the values for groups hashmap. Just need creation(bytes), value(generic) and Deletion/Empty (would replace NotFound). then validation is also the same as for values.